File: web/tests/reference/stacks/largest-rectangle.py
def largest_rectangle_area(heights):
    stack, best = [], 0
    for i, h in enumerate(heights + [0]):
        while stack and heights[stack[-1]] >= h:
            height = heights[stack.pop()]
            left = stack[-1] + 1 if stack else 0
            best = max(best, height * (i - left))
        stack.append(i)
    return best

File: web/tests/reference/stacks/largest-rectangle.js
function largestRectangleArea(heights) {
  const stack = []; let best = 0;
  const hs = [...heights, 0];
  for (let i = 0; i < hs.length; i++) {
    while (stack.length && hs[stack[stack.length - 1]] >= hs[i]) {
      const height = hs[stack.pop()];
      const left = stack.length ? stack[stack.length - 1] + 1 : 0;
      best = Math.max(best, height * (i - left));
    }
    stack.push(i);
  }
  return best;
}
